src/eddymotion/model/gradient_utils.py

```diff
@@ -0,0 +1,81 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+#
+# Copyright 2024 The NiPreps Developers <nipreps@gmail.com>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+import numpy as np
+
+from eddymotion.math.utils import compute_angle
+
+
+# ToDo
+# FSL crew takes as input dpars, which are diffusion parameters
+# Does this deal with multi-shell? What does the matrix look like cols vs rows?
+# Should the b0 be included or explicitly mention that it is excluded?
+# What if whe have different bvals ?
+def compute_pairwise_angles(bvecs, smallest):
+    r"""Compute pairwise angles across diffusion gradient encoding directions.
+
+    Following [Andersson15]_, it computes the smallest of the angles between
+    each pair if ``smallest`` is ``True``, i.e.
+
+    .. math::
+
+        \theta(\mathbf{g}, \mathbf{g'}) = \arccos(\abs{\langle \mathbf{g}, \mathbf{g'} \rangle})
+
+    Parameters
+    ----------
+    bvecs : :obj:`ndarray`
+        Diffusion gradient encoding directions in FSL format.
+    smallest : :obj:`bool`
+        ``True`` to compute the smallest angle between each direction pair.
+
+    Returns
+    -------
+    :obj:`ndarray`
+        Pairwise angles across diffusion gradient encoding directions.
+
+    References
+    ----------
+    .. [Andersson15] J. L. R. Andersson. et al., An integrated approach to
+       correction for off-resonance effects and subject movement in diffusion MR
+       imaging, NeuroImage 125 (2016) 1063–1078
+    """
+
+    if bvecs.shape[0] != 3:
+        raise ValueError(f"bvecs must be of shape (3, N). Found: {bvecs.shape}")
+
+    num_vectors = bvecs.shape[1]
+    angles = np.zeros((num_vectors, num_vectors))
+
+    # ToDo
+    # Note that the below also computes the angle of a given vector with itself:
+    # Should those be removed?
+    # ToDo
+    # Vectorize and refactor: bvecs.T @ bvecs to get the outer product, or use
```

Comment:
I agree with this:

- Vectors should be normalized before coming in, but nonetheless, it would not hurt to normalize here again (only once).
- We can move `compute_angle()` to be under tests and use it to validate this calculation if we obtain cosines matrix in one shot (bvecs.T @ bvecs) as you said here.

Reply:
> Vectors should be normalized before coming in, but nonetheless, it would not hurt to normalize here again (only once).

Done in https://github.com/nipreps/eddymotion/pull/200/files?diff=unified&w=0#diff-37c2c29d72b17bb336c5c519649dc18ff1f8c1ca0ddebde172b27ec208385935R59.

> We can move compute_angle() to be under tests and use it to validate this calculation if we obtain cosines matrix in one shot (bvecs.T @ bvecs) as you said here.

So used the inner product with the `@` operator and done as suggested 3fce5be. @arokem Oscar wanted to have it now, so done.

Removed all comments, including those about the multi-shell case, etc. We'll see how to deal with that when the time comes.